File: docs/nakama/guides/examples/league.lua
local nk = require("nakama")

-- Create the two tiers of leaderboards
local bottom_tier_id = "bottom-tier"
local top_tier_id = "top-tier"
local authoritative = true
local sort = "desc"
local operator = "best"
local reset = "0 0 * * 1"
local metadata = {}
nk.leaderboard_create(bottom_tier_id, authoritative, sort, operator, reset, metadata)
nk.leaderboard_create(top_tier_id, authoritative, sort, operator, reset, metadata)

-- Register leaderboard reset function to handle promotions and relegations
local function leaderboardReset(ctx, leaderboard, reset)
    -- We're only interested in our top/bottom tier leaderboards so return if it isn't them
    if leaderboard.id ~= bottom_tier_id and leaderboard.id ~= top_tier_id then
        return
    end

    -- Get all leaderboard records (assuming the tier has no more than 10,000 players)
    local records, _, _, _ = nk.leaderboard_records_list(leaderboard.id, {}, 10000)

    -- If leaderboard is top tier and has 10 or more players, relegate bottom 3 players
    if leaderboard.id == top_tier_id and #records >= 10 then
        -- Relegate record owner by copying their record into the bottom tier and deleting their current top tier record
        for i=3, 1, -1 do
            local record = records[#records-i]
            nk.leaderboard_record_write(bottom_tier_id, record.owner, record.username, record.score, record.subscore, {})
            nk.leaderboard_record_delete(top_tier_id, record.owner)
        end
    end

    -- If leaderboard is bottom tier and has 10 or more players, promote top 3 players
    if leaderboard.id == bottom_tier_id and #records >= 10 then
        -- Promote record owner by copying their record into the top tier and deleting their current top tier record
        for i=1, 3 do
            local record = records[i]
            nk.leaderboard_record_write(top_tier_id, record.owner, record.username, record.score, record.subscore, {})
            nk.leaderboard_record_delete(bottom_tier_id, record.owner)
        end
    end

    -- Distribute rewards based on player's tier
    for i=1, #records do
        local record = records[i]
        local reward = 100

        -- Increase reward for top tier players
        if leaderboard.id == top_tier_id then
            reward = 500
        end

        local changeset = {
            coins = reward
        }

        nk.wallet_update(record.owner, changeset, {}, true)
    end
end
nk.register_leaderboard_reset(leaderboardReset)

File: docs/nakama/guides/examples/league.ts
function InitModule(ctx: nkruntime.Context, logger: nkruntime.Logger, nk: nkruntime.Nakama, initializer: nkruntime.Initializer) {
    // Create the two tiers of leaderboards
    let bottomTierId = "bottom-tier";
    let topTierId = "top-tier";
    let authoritative = true;
    let sortOrder = nkruntime.SortOrder.DESCENDING;
    let operator = nkruntime.Operator.INCREMENTAL;
    let resetSchedule = "0 0 * * 1";
    let metadata = {};
    nk.leaderboardCreate(bottomTierId, authoritative, sortOrder, operator, resetSchedule, metadata);
    nk.leaderboardCreate(topTierId, authoritative, sortOrder, operator, resetSchedule, metadata);

    // Register leaderboard reset function to handle promotions and relegations
    let leaderboardReset: nkruntime.LeaderboardResetFunction = (ctx: nkruntime.Context, logger: nkruntime.Logger, nk: nkruntime.Nakama, leaderboard: nkruntime.Leaderboard, reset: number) => {
        // We're only interested in our top/bottom tier leaderboards so return if it isn't them
        if (leaderboard.id !== topTierId && leaderboard.id !== bottomTierId) {
            return;
        }

        let result = nk.leaderboardRecordsList(leaderboard.id, null, 10000, null, reset);

        // Get all leaderboard records (assuming the tier has no more than 10,000 players)
        if (leaderboard.id === topTierId && result.records.length >= 10) {
            // Relegate record owner by copying their record into the bottom tier and deleting their current top tier record
            result.records.slice(result.records.length - 3).forEach(r => {
                nk.leaderboardRecordWrite(bottomTierId, r.ownerId, r.username, r.score, r.subscore, null, null);
                nk.leaderboardRecordDelete(topTierId, r.ownerId);
            });
        }

        // If leaderboard is bottom tier and has 10 or more players, promote top 3 players
        if (leaderboard.id === topTierId && result.records.length >= 10) {
            // Promote record owner by copying their record into the top tier and deleting their current bottom tier record
            result.records.slice(0, 3).forEach(r => {
                nk.leaderboardRecordWrite(topTierId, r.ownerId, r.username, r.score, r.subscore, null, null);
                nk.leaderboardRecordDelete(bottomTierId, r.ownerId);
            });
        }

        // Distribute rewards based on player's tier
        result.records.forEach(r => {
            let reward = 100;

            // Increase reward for top tier players
            if (leaderboard.id === topTierId) {
                reward = 500;
            }

            let changeset = {
                coins: reward
            };

            nk.walletUpdate(r.ownerId, changeset, null, true);
        });
    };
    initializer.registerLeaderboardReset(leaderboardReset);
}

// Reference InitModule to avoid it getting removed on build
!InitModule && InitModule.bind(null);